File: go/2015/day02a/day02a.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func usage(progname string) {
	fmt.Fprintf(os.Stderr, "usage: %s <input file>\n", progname)
	os.Exit(1)
}

func process(filename string) uint32 {
	var totalArea uint32 = 0

	file, err := os.Open(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, "x")
		if len(parts) != 3 {
			fmt.Fprintf(os.Stderr, "Invalid line format: %s\n", line)
			os.Exit(1)
		}
		dimensions := make([]uint32, 3)
		for i, part := range parts {
			num, err := strconv.ParseUint(part, 10, 32)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Invalid number in line: %s", line)
				os.Exit(1)
			}
			dimensions[i] = uint32(num)
		}
		l, w, h := dimensions[0], dimensions[1], dimensions[2]
		area1 := l * w
		area2 := l * h
		area3 := w * h
		surfaceArea := 2 * area1 + 2 * area2 + 2 * area3
		minArea := area1
		if area2 < minArea {
			minArea = area2
		}
		if area3 < minArea {
			minArea = area3
		}
		totalArea += surfaceArea + minArea 
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	return totalArea
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}

	filename := os.Args[1]
	result := process(filename)
	fmt.Printf("result = %d\n", result)
}

File: c++/2015/day02a/day02a.cpp
#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

void usage(std::string progname) {
  std::cerr << "usage: " << progname << " <input file>" << std::endl;
  std::exit(1);
}

uint32_t process(std::string filename) {
  uint32_t total_area = 0;
  std::ifstream infile(filename);
  std::string line;
  while (std::getline(infile, line)) {
    std::stringstream ss(line);
    std::string token;
    std::vector<uint32_t> dimensions;
    while (std::getline(ss, token, 'x')) {
      uint32_t number = static_cast<uint32_t>(std::stoul(token));
      dimensions.push_back(number);
    }
    uint32_t area1 = dimensions[0] * dimensions[1];
    uint32_t area2 = dimensions[0] * dimensions[2];
    uint32_t area3 = dimensions[1] * dimensions[2];
    uint32_t surface_area = (area1 << 1) + (area2 << 1) + (area3 << 1);
    uint32_t min_area = std::min(std::min(area1, area2), area3);
    total_area += surface_area + min_area;
  }
  return total_area;
}

int main(int argc, char* argv[]) {
  std::string progname(argv[0]);
  if (argc < 2) {
    usage(progname);
  }
  std::string filename(argv[1]);
  int32_t result = process(filename);
  std::cout << "result = " << result << std::endl;
  return 0;
}
